product_restricted_type/models/product_template.py

```diff
@@ -1,27 +1,29 @@
-# Copyright 2018 ForgeFlow S.L.
+# Copyright 2018-2024 ForgeFlow S.L.
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
-from odoo import _, api, models
+from odoo import _, api, fields, models
 from odoo.exceptions import ValidationError
 
 
 class ProductTemplate(models.Model):
     _inherit = "product.template"
 
+    allowed_categ_ids = fields.Many2many(
+        "product.category",
+        compute="_compute_allowed_categ_ids",
+        string="Allowed Categories",
+    )
+
     @api.onchange("categ_id")
     def _onchange_categ_id(self):
         if self.categ_id and self.categ_id.restricted_product_type:
             self.type = self.categ_id.restricted_product_type
 
-    @api.onchange("type")
-    def _onchange_type(self):
-        if self.type:
-            return {
-                "domain": {
-                    "categ_id": [("restricted_product_type", "in", [self.type, False])]
-                }
-            }
-        else:
-            return {"domain": {"categ_id": []}}
+    @api.depends("type")
+    def _compute_allowed_categ_ids(self):
+        for record in self:
+            record.allowed_categ_ids = self.env["product.category"].search(
+                [("restricted_product_type", "in", [record.type, False])]
```

Comment:
Don't do a search in a loop.

Reply:
I've separated the search so that it is done before the actual assignation. Does it look better from your side?